controller/cmd/service-mirror/main.go

```diff
@@ -1,101 +1,160 @@
-	namespace := cmd.String("namespace", "", "address to serve scrapable metrics on")
+	namespace := cmd.String("namespace", "", "namespace containing Link and credentials Secret")
 	repairPeriod := cmd.Duration("endpoint-refresh-period", 1*time.Minute, "frequency to refresh endpoint resolution")
 
 	flags.ConfigureAndParse(cmd, args)
+	linkName := cmd.Arg(0)
 
 	stop := make(chan os.Signal, 1)
 	signal.Notify(stop, os.Interrupt, syscall.SIGTERM)
 
-	k8sAPI, err := k8s.InitializeAPI(
-		*kubeConfigPath,
-		false,
-		k8s.Svc,
-		k8s.NS,
-		k8s.Endpoint,
-	)
-
+	// We create two different kubernetes API clients for the local cluster:
+	// k8sAPI is used as a dynamic client for unstrcutured access to Link custom
+	// resources.
+	//
+	// controllerK8sAPI is used by the cluster watcher to manage
+	// mirror resources such as services, namespaces, and endpoints.
+	k8sAPI, err := k8s.NewAPI(*kubeConfigPath, "", "", []string{}, 0)
 	//TODO: Use can-i to check for required permissions
 	if err != nil {
 		log.Fatalf("Failed to initialize K8s API: %s", err)
 	}
 
-	secretsInformer, err := initLocalResourceInformer(k8sAPI.Client, *namespace, k8s.Secret)
+	controllerK8sAPI, err := controllerK8s.InitializeAPI(*kubeConfigPath, false,
+		controllerK8s.NS,
+		controllerK8s.Svc,
+		controllerK8s.Endpoint,
+	)
 	if err != nil {
-		log.Fatalf("Failed to initialize secret informer: %s", err)
+		log.Fatalf("Failed to initialize K8s API: %s", err)
 	}
-	svcInformer, err := initLocalResourceInformer(k8sAPI.Client, *namespace, k8s.Svc)
 
+	gvr := schema.GroupVersionResource{
+		Group:    "multicluster.linkerd.io",
+		Version:  "v1alpha1",
+		Resource: "links",
+	}
+	linkClient := k8sAPI.DynamicClient.Resource(gvr).Namespace(*namespace)
+
+	go admin.StartServer(*metricsAddr)
+
+	controllerK8sAPI.Sync(nil)
+
+	// Start link watch
+	linkWatch, err := linkClient.Watch(metav1.ListOptions{})
 	if err != nil {
-		log.Fatalf("Failed to initialize service informer: %s", err)
+		log.Fatalf("Failed to watch Link %s: %s", linkName, err)
+	}
+	results := linkWatch.ResultChan()
+
+	// Each time the link resource is updated, reload the config and restart the
+	// cluster watcher.
+	for event := range results {
+		switch obj := event.Object.(type) {
+		case *dynamic.Unstructured:
+			if obj.GetName() == linkName {
+				switch event.Type {
+				case watch.Added, watch.Modified:
+					link, err := multicluster.NewLink(*obj)
+					if err != nil {
+						log.Errorf("Failed to parse link %s: %s", linkName, err)
+						continue
+					}
+					log.Infof("Got updated link %s: %+v", linkName, link)
+					creds, err := loadCredentials(link, *namespace, k8sAPI)
+					if err != nil {
+						log.Errorf("Failed to load remote cluster credentials: %s", err)
+					}
+					restartClusterWatcher(link, *namespace, creds, controllerK8sAPI, *requeueLimit, *repairPeriod)
+				case watch.Deleted:
+					log.Infof("Link %s deleted", linkName)
+					// TODO: should we delete all mirror resources?
```

Comment:
I would assume so

Reply:
If we can tie all mirror services to this link why wouldn't we?